pybop/optimisers/pints_optimisers.py

```diff
@@ -1,8 +1,428 @@
 import numpy as np
 import pints
 
+from .base_optimiser import BaseOptimiser
 
-class GradientDescent(pints.GradientDescent):
+
+class BasePintsOptimiser(BaseOptimiser):
+    """
+    A base class for defining optimisation methods from the PINTS library.
+
+    Parameters
+    ----------
+    x0 : array_like
+        Initial position from which optimization will start.
+    sigma0 : float, optional
+        Initial step size or standard deviation depending on the optimiser (default is 0.1).
+    bounds : dict, optional
+        A dictionary with 'lower' and 'upper' keys containing arrays for lower and upper
+        bounds on the parameters.
+    **optimiser_kwargs : optional
+        Valid PINTS option keys and their values.
+    """
+
+    def __init__(
+        self, pints_class, x0=None, sigma0=None, bounds=None, **optimiser_kwargs
+    ):
+        super().__init__(x0, sigma0, bounds)
+
+        # Convert bounds to PINTS boundaries
+        if bounds is not None:
+            boundaries = pints.RectangularBoundaries(bounds["lower"], bounds["upper"])
+        else:
+            boundaries = None
+        self._boundaries = boundaries
+
+        # Convert x0 to PINTS vector
+        self._x0 = pints.vector(self.x0)
+
+        # PyBOP doesn't currently support the PINTS transformation class
+        self._transformation = None
+
+        # Create an instance of the PINTS optimiser class
+        self.pints_optimiser = pints_class(x0, sigma0, self._boundaries)
+
+        # Check if sensitivities are required
+        self._needs_sensitivities = self.pints_optimiser.needs_sensitivities()
+
+        # Track optimiser's f_best or f_guessed
+        self._use_f_guessed = None
+        self.set_f_guessed_tracking()
+
+        # Parallelisation
+        self._parallel = False
+        self._n_workers = 1
+        self.set_parallel()
+
+        # User callback
+        self._callback = None
+
+        # Define stopping criteria
+        # Maximum iterations set in BaseOptimiser
+
+        # Minimum iterations
+        self._min_iterations = None
+        self.set_min_iterations()
+
+        # Maximum unchanged iterations
+        self._unchanged_threshold = 1  # smallest significant f change
+        self._unchanged_max_iterations = None
+        self.set_max_unchanged_iterations()
+
+        # Maximum evaluations
+        self._max_evaluations = None
+
+        # Threshold value
+        self._threshold = None
+
+        # Post-run statistics
+        self._evaluations = None
+        self._iterations = None
+
+        self.update_options(**optimiser_kwargs)
+
+    def update_options(self, **optimiser_kwargs):
+        """
+        Update the optimiser options.
+
+        Parameters
+        ----------
+        **optimiser_kwargs : optional
+            Valid PINTS option keys and their values.
+        """
+        for key, value in optimiser_kwargs.items():
```

Comment:
I wonder if there is a better way to map the kwarg to the individual option. This is feels like it could get overwhelming if we start adding additional options.

Reply:
Outdated